src/program.hpp

```diff
@@ -395,15 +396,15 @@ struct cvk_program : public _cl_program, api_object {
     VkShaderModule shader_module() const { return m_shader_module; }
 
     void wait_for_operation() {
+        CVK_ASSERT(m_thread->joinable());
         m_thread->join();
-        delete m_thread;
     }
 
     void complete_operation(cvk_device* device, cl_build_status status) {
-        m_dev_status[device] = status;
         if (m_operation_callback != nullptr) {
             m_operation_callback(this, m_operation_callback_data);
         }
+        m_dev_status[device] = status;
```

Comment:
This doesn't make sense to me. It essentially means that the callback can't observe the status of the build which feels wrong. In general, the spirit of the spec wrt. callbacks is that they are called at some undefined point after a given event. I think the test should probably loop until it confirms the callback has been called, probably with a max number of attempts so test doesn't deadlock on failure. I think we should probably clarify the spec as well and add two tests where he callback is required to observe a successful and failed build.

Reply:
Yeah I agree this did feel weird. I opted for this approach to avoid making changes to CTS that might break existing implementations, but I'm also fine with clarifying the spec and modifying CTS.